### PR TITLE
Close mongodb cursor if the iterator has no more items

### PR DESCRIPTION
**Pull request by saxenakshitiz**

### CODE REVIEW

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java [hunk 1]
```diff
@@ -582,7 +582,11 @@ private BasicDBObject selectionCriteriaForKeys(Set<Key> keys) {
     return new Iterator<>() {
       @Override
       public boolean hasNext() {
-        return cursor.hasNext();
+        boolean hasNext = cursor.hasNext();
+        if (!hasNext) {
+          cursor.close();
+        }
+        return hasNext;
```
Comment (avinashkolluru) [inline document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java:589]: How do we handle cases where next() throws an error. Wouldn't the cursor remain open in those cases

Comment (aaron-steinfeld) [reply document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java:589]: We should probably catch in that case. Would it be feasible to additionally make this a closable iterator? Basically what MongoCursor is doing. The whole design of doc store is problematic in that it returns an iterator which is left open until completely iterated - this fix only resolves the case where we complete iteration.

Comment (avinashkolluru) [reply document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java:589]: Yeah. A closeable iterator is a better option. But that would mean its the consumers responsibility to close now.

Comment (aaron-steinfeld) [reply document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java:589]: I was thinking both - closing it on completion like this PR is doing, but aso returning something that maintains compatibility with the existing API like
```java
interface Cursor<TResult> extends Iterator<TResult>, Closeable {}
```

Comment (saxenakshitiz) [reply document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java:597]: Made this change

File: document-store/src/main/java/org/hypertrace/core/documentstore/ClosableIterator.java [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package org.hypertrace.core.documentstore;
+
+import java.io.Closeable;
+import java.util.Iterator;
+
+public interface ClosableIterator<E> extends Iterator<E>, Closeable {}
```
Comment (aaron-steinfeld) [inline document-store/src/main/java/org/hypertrace/core/documentstore/ClosableIterator.java:6]:
```suggestion
public interface CloseableIterator<E> extends Iterator<E>, Closeable {}
```
Should match jdk interface name

**Review by aaron-steinfeld — APPROVED** [review]

**Review by aaron-steinfeld — APPROVED** [review]

File: document-store/src/main/java/org/hypertrace/core/documentstore/ClosableIterator.java (hunk 2 above)
Comment (saxenakshitiz) [reply document-store/src/main/java/org/hypertrace/core/documentstore/ClosableIterator.java]: Done

**Pull request merged**